File: leetcode.com/0838_push_dominoes/solution.py
class Solution:
    def pushDominoes(self, dominoes: str) -> str:

        def fill(l, r, ch):
            for i in range(l, r + 1):
                arr[i] = ch

        arr = list(dominoes)
        l = 0
        r = None

        for i, ch in enumerate(arr):
            if ch == 'L' and r is None:
                fill(l, i, ch)
            elif ch == 'L' and r is not None:
                mid = (i - r - 1) // 2
                fill(r + 1, r + mid, 'R')
                fill(i - mid, i - 1, 'L')
                r = None
                l = i
            elif ch == 'R':
                if r is not None:
                    fill(r, i, 'R')
                r = i

        if r is not None:
            fill(r, len(arr) - 1, 'R')

        return ''.join(arr)

File: leetcode.com/0838_push_dominoes/solution.c
inline void fill(char *a, int l, int r, char ch)
{
    for (auto int j = l ; j <= r ; j++) a[j] = ch;
}

char *pushDominoes(char *dominoes)
{
    int len = strlen(dominoes);
    int l = 0;
    int r = -1;

    for (auto int i = 0 ; i < len ; i++) {
        if (dominoes[i] == '.')
            continue;

        if (dominoes[i] == 'L') {
            if (r == -1) {
                fill(dominoes, l, i, 'L');
            } else {
                int mid = (i - r - 1) / 2;
                fill(dominoes, r + 1, r + mid, 'R');
                fill(dominoes, i - mid, i - 1, 'L');
                r = -1;
                l = i;
            }
        } else if (dominoes[i] == 'R') {
            if (r != -1)
                fill(dominoes, r, i, 'R');
            r = i;
        }
    }

    if (r != -1)
        fill(dominoes, r, len - 1, 'R');

    return dominoes;
}
